File: pdfbridge/superposer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (C) 2014 The ProteinDF development team.
# see also AUTHORS and README if provided.
# 
# This file is a part of the ProteinDF software package.
# 
# The ProteinDF is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# The ProteinDF is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with ProteinDF.  If not, see <http://www.gnu.org/licenses/>.

import math
import copy

import pdfbridge

class Superposer(object):
    """
    """

    def __init__(self, atom_group1, atom_group2):
        """
        @param atom_group1[in] RMSD・回転行列を求めたい分子団
        @param atom_group2[in] 基準となる分子団
        """
        self._atomgroup1 = pdfbridge.AtomGroup(atom_group1)
        self._atomgroup2 = pdfbridge.AtomGroup(atom_group2)
        (positions1, positions2) = self._match_positions(atom_group1, atom_group2)
        self._num_of_positions = len(positions1)
        assert(self._num_of_positions == len(positions2))
        self._positions1 = positions1
        self._positions2 = positions2

        self._center1 = None
        self._center2 = None
        self._shift_positions1 = None
        self._shift_positions2 = None
        self._rotation_mat = None
        self._update_positions1 = None
        self._update_positions2 = None
        self._rmsd = None

        #self._calc(self._atomgroup1,
        #           self._atomgroup2)

    #
    def _get_num_of_positions(self):
      return self._num_of_positions

    num_of_positions = property(_get_num_of_positions)
    # -----------------------------------------------------------------
    def _get_positions1(self):
        return self._positions1

    positions1 = property(_get_positions1)
    # -----------------------------------------------------------------
    def _get_positions2(self):
        return self._positions2

    positions2 = property(_get_positions2)
    # -----------------------------------------------------------------
    def _get_center1(self):
        if self._center1 == None:
            self._center1 = self.get_center(self.positions1)
            #print("center1: {}".format(self._center1))
        return self._center1

    center1 = property(_get_center1)
    # -----------------------------------------------------------------
    def _get_center2(self):
        if self._center2 == None:
            self._center2 = self.get_center(self.positions2)
            #print("center2: {}".format(self._center2))
        return self._center2

    center2 = property(_get_center2)
    # -----------------------------------------------------------------
    def _get_shift_positions1(self):
        if self._shift_positions1 == None:
            self._shift_positions1 = self._shift_positions(self.positions1, self.center1)
            #print(self._shift_positions1)
        return self._shift_positions1

    shift_positions1 = property(_get_shift_positions1)
    # -----------------------------------------------------------------
    def _get_shift_positions2(self):
        if self._shift_positions2 == None:
            self._shift_positions2 = self._shift_positions(self.positions2, self.center2)
            #print(self._shift_positions2)
        return self._shift_positions2

    shift_positions2 = property(_get_shift_positions2)
    # -----------------------------------------------------------------
    def _get_rotation_mat(self):
        if self._rotation_mat == None:
            self._rotation_mat = self._get_rotation_matrix(self.num_of_positions,
                                                           self.shift_positions1,
                                                           self.shift_positions2)
        return self._rotation_mat

    rotation_mat = property(_get_rotation_mat)
    # -----------------------------------------------------------------
    def _get_update_positions1(self):
        # positions1 を回転
        if self._update_positions1 == None:
            self._update_positions1 = copy.deepcopy(self.shift_positions1)
            for i in range(len(self.positions1)):
                self._update_positions1[i].rotate(self.rotation_mat)
        return self._update_positions1

    update_positions1 = property(_get_update_positions1)
    # -----------------------------------------------------------------
    def _get_update_positions2(self):
        # positions2 はそのまま
        if self._update_positions2 == None:
            self._update_positions2 = copy.deepcopy(self.shift_positions2)
        return self._update_positions2

    update_positions2 = property(_get_update_positions2)
    # -----------------------------------------------------------------
    def _get_rmsd(self):
        if self._rmsd == None:
            self._rmsd = self._calc_rmsd(self.update_positions1,
                                         self.update_positions2)
        return self._rmsd

    rmsd = property(_get_rmsd)
    # -----------------------------------------------------------------
    def superimpose(self, atomgroup):
        # positions1 を移動
        answer = pdfbridge.AtomGroup(atomgroup)
        answer.shift_by(- self.center1)

        answer.rotate(self.rotation_mat)

        # positions2 のcenterに移動
        answer.shift_by(self.center2)
        
        return answer
        
    # -----------------------------------------------------------------
    def _calc(self, atom_group1, atom_group2):
        (positions1, positions2) = self._match_positions(atom_group1, atom_group2)
        num_of_positions = len(positions1)
        assert(num_of_positions == len(positions2))

        (translation_vct1, self._translation_vct2) = self._fix_positions(positions1, positions2)

        self._rotation_mat = self._get_rotation_matrix(num_of_positions,
                                                       positions1, positions2)

        self._update_positions(positions1, positions2,
                               self._rotation_mat,
                               self._translation_vct2)

        self._rmsd = self._calc_rmsd(positions1, positions2)
        #print('<<<< superposer')

    def _match_positions(self, atom_group1, atom_group2):
        """
        compare two atom_group objects, and select common points.
        return the list of list corresponding two points.
        """
        positions1 = []
        positions2 = []
        for key, ag1 in atom_group1.groups():
            if atom_group2.has_group(key):
                (p1, p2) = self._match_positions(ag1,
                                                 atom_group2.get_group(key))
                positions1 += p1
                positions2 += p2

        for key, atom1 in atom_group1.atoms():
            if atom_group2.has_atom(key):
                #print(str(atom1), str(atom_group2.get_atom(key)))
                positions1 += [atom1.xyz]
                positions2 += [atom_group2.get_atom(key).xyz]

        return (positions1, positions2)

    def _fix_positions(self, positions1, positions2):
        center1 = self.get_center(positions1)
        center2 = self.get_center(positions2)

        translation_vct1 = - center1
        translation_vct2 =   center2

        for i in range(len(positions1)):
            positions1[i] -= center1
        for i in range(len(positions2)):
            positions2[i] -= center2

        return (translation_vct1, translation_vct2)


    def get_center(self, positions):
        """
        return the center position of input positions
        """
        c = pdfbridge.Position()

        num_of_positions = len(positions)
        for i in range(num_of_positions):
            c += positions[i]
        c /= num_of_positions

        return c

    def _shift_positions(self, positions, center):
        answer = [ p - center for p in positions ]

        # check
        sum_of_positions = pdfbridge.Position()
        for p in answer:
            sum_of_positions += p
        assert(sum_of_positions.distance_from() < 1.0E-5)

        return answer

    def _get_rotation_matrix(self, num_of_points,
                             positions1, positions2):
        r = pdfbridge.Matrix(3, 3)

        # r_ij = Sum_over_k{p2(k, i) * p1(k, j)}
        for k in range(num_of_points):
            x1 = positions1[k].x
            y1 = positions1[k].y
            z1 = positions1[k].z
            x2 = positions2[k].x
            y2 = positions2[k].y
            z2 = positions2[k].z

            r.add(0, 0, x2 * x1)
            r.add(0, 1, x2 * y1)
            r.add(0, 2, x2 * z1)
            r.add(1, 0, y2 * x1)
            r.add(1, 1, y2 * y1)
            r.add(1, 2, y2 * z1)
            r.add(2, 0, z2 * x1)
            r.add(2, 1, z2 * y1)
            r.add(2, 2, z2 * z1)

        #print(' > rot mat: r')
        #print(r)
        tr = r.copy()
        tr.transpose()
        #print(' > rot mat: tr')
        #print(tr)
        trr = tr * r
        #print(' > rot mat: trr')
        #print(trr)
        trr = trr.get_symmetric_matrix()
        #print(trr)

        eigval, eigvec = trr.eig()
        #print('eigval')
        #print(eigval)
        #print('eigvec')
        #print(eigvec)
        a = self._make_right_handed(eigvec)
        #print('a')
        #print(a)

        b = pdfbridge.Matrix(3, 3)
        for i in range(3):
            for j in range(3):
                for k in range(3):
                    v = r.get(j , k) * a.get(i, k)
                    b.add(i, j, v)

            # normalize b[i]
            w = 0.0
            for j in range(3):
                w += b.get(i, j) * b.get(i, j)

            t = math.sqrt(1.0 / w)
            for j in range(3):
                v = b.get(i, j)
                b.set(i, j, v * t)
        #print('b')
        #print(b)

        # b[2] = b[0] x b[1]
        tmp_vct = self._calc_vector_product(b.get_row_vector(0),
                                            b.get_row_vector(1))
        for i in range(3):
            b.set(2, i, tmp_vct[i])

        # rotation matrix r_ij = b_ki * a_kj
        mat = self._set_rotation(a, b)
        return mat

    def _make_right_handed(self, mat):
        assert(mat.rows == 3)
        assert(mat.cols == 3)

        v1 = pdfbridge.Vector(3)
        v2 = pdfbridge.Vector(3)
        for i in range(3):
            v1[i] = mat.get(0, i)
            v2[i] = mat.get(1, i)

        v3 = self._calc_vector_product(v1, v2)

        answer = pdfbridge.Matrix(3, 3)
        for i in range(3):
            answer.set(0, i, v1[i])
            answer.set(1, i, v2[i])
            answer.set(2, i, v3[i])

        return answer

    def _calc_vector_product(self, v1, v2):
        assert(len(v1) == 3)
        assert(len(v2) == 3)

        v3 = pdfbridge.Vector(3)
        v3[0] = v1[1] * v2[2] - v1[2] * v2[1]
        v3[1] = v1[2] * v2[0] - v1[0] * v2[2]
        v3[2] = v1[0] * v2[1] - v1[1] * v2[0]

        return v3

    def _set_rotation(self, a, b):
        assert(a.rows == 3)
        assert(a.cols == 3)
        assert(b.rows == 3)
        assert(b.cols == 3)

        r = pdfbridge.Matrix(3, 3)
        for i in range(3):
            for j in range(3):
                for k in range(3):
                    v = b.get(k, i) * a.get(k, j)
                    r.add(i, j, v)
        return r

    def _update_positions(self, positions1, positions2,
                          rotation_mat, translation_vct2):
        for i in range(len(positions1)):
            positions1[i].rotate(rotation_mat)

        for i in range(len(positions1)):
            positions1[i] += translation_vct2
        for i in range(len(positions2)):
            positions2[i] += translation_vct2

    def _calc_rmsd(self, positions1, positions2):
        """
        calc rmsd.
        store the value to self._rmsd
        """
        num_of_positions = min(len(self.update_positions1),
                               len(self.update_positions2))
        msd = 0.0
        for i in range(num_of_positions):
            msd += positions1[i].square_distance_from(positions2[i])
        msd /= float(num_of_positions)
        rmsd = math.sqrt(msd)

        return rmsd

    def _debug_positions(self, positions):
        output = ""
        for i, p in enumerate(positions):
            output += "[{}] {}\n".format(i, str(p))
        return output

File: rust/crates/proteindf-bridge/src/superposer.rs
// SPDX-FileCopyrightText: The ProteinDF development team
// SPDX-License-Identifier: GPL-3.0-or-later

use crate::atom_group::AtomGroup;
use crate::error::{BridgeError, Result};
use crate::matrix::Matrix;
use crate::position::Position;
use crate::vector::Vector;

/// Structure for superposing two AtomGroups using the Kabsch algorithm.
///
/// Ports `proteindf_bridge.superposer.Superposer`.
#[derive(Debug, Clone)]
pub struct Superposer {
    atom_group1: AtomGroup,
    atom_group2: AtomGroup,
    num_of_positions: usize,
    positions1: Vec<Position>,
    positions2: Vec<Position>,
    center1: Position,
    center2: Position,
    shift_positions1: Vec<Position>,
    shift_positions2: Vec<Position>,
    rotation_mat: Matrix,
    update_positions1: Vec<Position>,
    update_positions2: Vec<Position>,
    rmsd: f64,
}

impl Superposer {
    /// Creates a new `Superposer` from two `AtomGroup` instances and computes RMSD and rotation matrix.
    pub fn new(atom_group1: &AtomGroup, atom_group2: &AtomGroup) -> Result<Self> {
        let (positions1, positions2) = Self::match_positions(atom_group1, atom_group2);
        let num_of_positions = positions1.len();
        if num_of_positions != positions2.len() {
            return Err(BridgeError::value_error(
                "positions",
                "Matched positions between atom groups must have equal length",
            ));
        }
        if num_of_positions == 0 {
            return Err(BridgeError::value_error(
                "positions",
                "No common atom positions found to superpose",
            ));
        }

        let center1 = Self::calc_center(&positions1);
        let center2 = Self::calc_center(&positions2);

        let shift_positions1 = Self::shift_positions(&positions1, center1);
        let shift_positions2 = Self::shift_positions(&positions2, center2);

        let rotation_mat =
            Self::calc_rotation_matrix(num_of_positions, &shift_positions1, &shift_positions2)?;

        // Rotate update_positions1 by rotation_mat
        let mut update_positions1 = shift_positions1.clone();
        for pos in &mut update_positions1 {
            pos.rotate(&rotation_mat)?;
        }

        let update_positions2 = shift_positions2.clone();

        let rmsd = Self::calc_rmsd(&update_positions1, &update_positions2);

        Ok(Self {
            atom_group1: atom_group1.clone(),
            atom_group2: atom_group2.clone(),
            num_of_positions,
            positions1,
            positions2,
            center1,
            center2,
            shift_positions1,
            shift_positions2,
            rotation_mat,
            update_positions1,
            update_positions2,
            rmsd,
        })
    }

    /// Returns a reference to the first atom group.
    pub fn atom_group1(&self) -> &AtomGroup {
        &self.atom_group1
    }

    /// Returns a reference to the second atom group.
    pub fn atom_group2(&self) -> &AtomGroup {
        &self.atom_group2
    }

    /// Number of matched common points between the two atom groups.
    pub fn num_of_positions(&self) -> usize {
        self.num_of_positions
    }

    /// First atom group's matched positions.
    pub fn positions1(&self) -> &[Position] {
        &self.positions1
    }

    /// Second atom group's matched positions.
    pub fn positions2(&self) -> &[Position] {
        &self.positions2
    }

    /// Centroid of positions in the first atom group.
    pub fn center1(&self) -> Position {
        self.center1
    }

    /// Centroid of positions in the second atom group.
    pub fn center2(&self) -> Position {
        self.center2
    }

    /// Centroid-shifted positions of the first atom group.
    pub fn shift_positions1(&self) -> &[Position] {
        &self.shift_positions1
    }

    /// Centroid-shifted positions of the second atom group.
    pub fn shift_positions2(&self) -> &[Position] {
        &self.shift_positions2
    }

    /// 3x3 rotation matrix calculated to minimize RMSD.
    pub fn rotation_mat(&self) -> &Matrix {
        &self.rotation_mat
    }

    /// Rotated and shifted positions of the first atom group.
    pub fn update_positions1(&self) -> &[Position] {
        &self.update_positions1
    }

    /// Shifted positions of the second atom group (reference).
    pub fn update_positions2(&self) -> &[Position] {
        &self.update_positions2
    }

    /// Root-mean-square deviation (RMSD) between the two groups.
    pub fn rmsd(&self) -> f64 {
        self.rmsd
    }

    /// Superimposes the given `atomgroup` by shifting by `-center1`, rotating by `rotation_mat`,
    /// and shifting by `+center2`.
    pub fn superimpose(&self, atomgroup: &AtomGroup) -> Result<AtomGroup> {
        let mut answer = atomgroup.clone();
        answer.shift_by(-self.center1);
        answer.rotate(&self.rotation_mat)?;
        answer.shift_by(self.center2);
        Ok(answer)
    }

    /// Recursively matches common atom positions between two atom groups.
    fn match_positions(
        atom_group1: &AtomGroup,
        atom_group2: &AtomGroup,
    ) -> (Vec<Position>, Vec<Position>) {
        let mut positions1 = Vec::new();
        let mut positions2 = Vec::new();

        for (key, ag1) in atom_group1.groups() {
            if let Some(ag2) = atom_group2.get_group(key) {
                let (p1, p2) = Self::match_positions(ag1, ag2);
                positions1.extend(p1);
                positions2.extend(p2);
            }
        }

        for (key, atom1) in atom_group1.atoms() {
            if let Some(atom2) = atom_group2.get_atom(key) {
                positions1.push(atom1.xyz);
                positions2.push(atom2.xyz);
            }
        }

        (positions1, positions2)
    }

    /// Computes the geometric center (centroid) of positions.
    fn calc_center(positions: &[Position]) -> Position {
        if positions.is_empty() {
            return Position::default();
        }
        let mut sum = Position::default();
        for p in positions {
            sum += *p;
        }
        sum / (positions.len() as f64)
    }

    /// Shifts all positions by subtracting `center`.
    fn shift_positions(positions: &[Position], center: Position) -> Vec<Position> {
        positions.iter().map(|p| *p - center).collect()
    }

    /// Computes the optimal 3x3 rotation matrix using the Kabsch algorithm.
    fn calc_rotation_matrix(
        num_of_points: usize,
        positions1: &[Position],
        positions2: &[Position],
    ) -> Result<Matrix> {
        let mut r = Matrix::new(3, 3);

        // r_ij = Sum_over_k { p2(k, i) * p1(k, j) }
        for k in 0..num_of_points {
            let x1 = positions1[k].x;
            let y1 = positions1[k].y;
            let z1 = positions1[k].z;
            let x2 = positions2[k].x;
            let y2 = positions2[k].y;
            let z2 = positions2[k].z;

            r.add(0, 0, x2 * x1);
            r.add(0, 1, x2 * y1);
            r.add(0, 2, x2 * z1);
            r.add(1, 0, y2 * x1);
            r.add(1, 1, y2 * y1);
            r.add(1, 2, y2 * z1);
            r.add(2, 0, z2 * x1);
            r.add(2, 1, z2 * y1);
            r.add(2, 2, z2 * z1);
        }

        let tr = r.transpose();

        let trr = &tr * &r;
        let trr_sym = trr.get_symmetric_matrix()?;

        let (_eigval, eigvec) = trr_sym.eig()?;

        // NOTE: In the original Python implementation:
        // `a = self._make_right_handed(eigvec)` is called with `eigvec`.
        let a = Self::make_right_handed(&eigvec)?;

        let mut b = Matrix::new(3, 3);
        for i in 0..3 {
            for j in 0..3 {
                for k in 0..3 {
                    let v = r.get(j, k)? * a.get(i, k)?;
                    b.add(i, j, v);
                }
            }

            // Normalize row b[i]
            let mut w = 0.0;
            for j in 0..3 {
                let v = b.get(i, j)?;
                w += v * v;
            }

            if w > 0.0 {
                let t = (1.0 / w).sqrt();
                for j in 0..3 {
                    let v = b.get(i, j)?;
                    b.set(i, j, v * t);
                }
            }
        }

        // rotation matrix r_ij = b_ki * a_kj
        Self::set_rotation(&a, &b)
    }

    /// Makes the coordinate system right-handed.
    fn make_right_handed(mat: &Matrix) -> Result<Matrix> {
        let mut v1 = Vector::new(3);
        let mut v2 = Vector::new(3);
        for i in 0..3 {
            v1.set(i, mat.get(0, i)?);
            v2.set(i, mat.get(1, i)?);
        }

        let v3 = Self::calc_vector_product(&v1, &v2)?;

        let mut answer = Matrix::new(3, 3);
        for i in 0..3 {
            answer.set(0, i, v1.get(i)?);
            answer.set(1, i, v2.get(i)?);
            answer.set(2, i, v3.get(i)?);
        }

        Ok(answer)
    }

    /// Computes the cross product of two 3D vectors.
    fn calc_vector_product(v1: &Vector, v2: &Vector) -> Result<Vector> {
        let mut v3 = Vector::new(3);
        v3.set(0, v1.get(1)? * v2.get(2)? - v1.get(2)? * v2.get(1)?);
        v3.set(1, v1.get(2)? * v2.get(0)? - v1.get(0)? * v2.get(2)?);
        v3.set(2, v1.get(0)? * v2.get(1)? - v1.get(1)? * v2.get(0)?);
        Ok(v3)
    }

    /// Computes rotation matrix R_ij = sum_k { b_ki * a_kj }.
    fn set_rotation(a: &Matrix, b: &Matrix) -> Result<Matrix> {
        let mut r = Matrix::new(3, 3);
        for i in 0..3 {
            for j in 0..3 {
                for k in 0..3 {
                    let v = b.get(k, i)? * a.get(k, j)?;
                    r.add(i, j, v);
                }
            }
        }
        Ok(r)
    }

    /// Computes RMSD between two sets of positions.
    fn calc_rmsd(positions1: &[Position], positions2: &[Position]) -> f64 {
        let n = positions1.len().min(positions2.len());
        if n == 0 {
            return 0.0;
        }
        let mut msd = 0.0;
        for i in 0..n {
            msd += positions1[i].square_distance_from(&positions2[i]);
        }
        (msd / (n as f64)).sqrt()
    }
}
